File: 0442-find-all-duplicates-in-an-array/0442-find-all-duplicates-in-an-array.cpp
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans;
        
        if (n < 2) return ans;
        
        for (int i=0; i<n; ++i) {
            int index = abs(nums[i]) - 1;
            if (nums[index] < 0) ans.push_back(index + 1);
            else nums[index] = -nums[index];
        }
        for (int i=0; i<n; ++i)
            nums[i] = abs(nums[i]);
        
        return ans;
    }
};

File: 0442-find-all-duplicates-in-an-array/0442-find-all-duplicates-in-an-array.java
// class Solution {
//     public List<Integer> findDuplicates(int[] numRay) {
//         List<Integer> l = new ArrayList<>();
//         if(numRay.length < 2) return l;
        
//         for (int i = 0; i < numRay.length; i++) {
//             numRay[(numRay[i] - 1) % numRay.length] += numRay.length;
//         }
        
//         for (int i = 0; i < numRay.length; i++) {
//             if (numRay[i] >= (2 * numRay.length)) 
//                 l.add(i + 1);
//         }
        
//         return l;
//     }
// }

class Solution {
    public List<Integer> findDuplicates(int[] numRay) {
        List<Integer> l = new ArrayList<>();
        if (numRay.length < 2) return l;

        for (int i = 0; i < numRay.length; i++) {
            int index = Math.abs(numRay[i]) - 1;
            if (numRay[index] < 0) {
                l.add(index + 1);
            } else {
                numRay[index] = -numRay[index];
            }
        }

        for (int i = 0; i < numRay.length; i++) {
            numRay[i] = Math.abs(numRay[i]);
        }

        return l;
    }
}
